### PR TITLE
oneapi: refactoring and rpaths

### PR DESCRIPTION
**Pull request by skosukhin**
The two main features of this PR are:
1. Generate `.cfg` files for the compilers that contain rpath flags for the directories with the runtime libraries. This way, executables that are generated by the compilers contain rpaths to the libraries they need (e.g. `libimf.so`, `libiomp5.so`, `libintlc.so`, etc.) and there is no need to load modules or set `LD_LIBRARY_PATH` to be able to run the executables. We do something similar for `gcc` (see [here](https://github.com/spack/spack/blob/1bd33d88bd5bdf6820800b71b2f5f798e857920b/var/spack/repos/builtin/packages/gcc/package.py#L753)).
2. Extend the module file generated for `intel-oneapi-mkl` so that an executable produced with `-mkl/-qmkl` Intel compiler flag contains rpaths to the MKL libraries.

The rest is refactoring. I tried to keep the history clean and the commit messages are hopefully informative enough (yes, I should have probably created a separate PR with the refactoring).

I also wanted to discuss one thing here. I don't remember where exactly, probably on a slide of a presentation about Spack, but there was one bullet point on the list of what Spack offers: "user can use environment modules but they do not have to". I try to follow and support that as much as possible. This is, for example, why I think we should generate the `.cfg` files (see above) under the default names and not under some other ones and set `ICXCFG` (and friends) in the module files. However, `intel-oneapi-compilers` are not there yet. For example, we get a failure when we call one of the compilers with the `-mkl/-qmkl` flag without setting `MKLROOT` or loading the `intel-oneapi-mkl` module first. By the way, I would not see any problem if the compilers didn't have the `-mkl/-qmkl` flag at all: there would not be any magic promised by the compiler and no magic would be expected by the users. However, the flags are there and we basically get a broken installation of Intel compilers. As far as I understand, the problem here is that we contradict the way intended by the compiler developers by following the Spack approach of installing everything to different prefixes. I don't mind that but I would like Spack to be able to produce a fully functional installation of the compiler toolchain. In particular, I would like the `-mkl/-qmkl` flags to work without any modules or additional environment variables that the users have to load/set. I see several possible solutions to that:
1. Introduce additional variants for `intel-oneapi-compilers` like `+mkl`, `+mpi`, etc. and install the respective components to the same prefix. This is probably the closest to how the compiler developers see it. The problem here is that we get duplicates in the installation tree: `intel-oneapi-mkl`, which we might want to install to use with other compilers, will be a subset of `intel-oneapi-compilers +mkl`.
2. Make `intel-oneapi-mkl` extend `intel-oneapi-compilers`, which would allow for `spack activate intel-oneapi-mkl. Two problems here: 
    a) according to how the extension are currently implemented in Spack, `intel-oneapi-compilers` becomes a dependency of `intel-oneapi-mkl`, which we don't really want since the user might want to install the MKL library only, without the compilers;
    b) we have to ensure correct file conflict resolution: there are files with the same names in the prefixes of `intel-oneapi-compilers` and `intel-oneapi-mkl`, which breaks the activation as it is currently implemented in Spack.
3. Introduce a meta-package `intel-oneapi-toolchain`, which would have variants `+mkl`, `+mpi`, etc,  have the respective components as conditional dependencies and install nothing but create a view (with symlinks) of all of the components in its prefix.

Different hybrid solutions of the three above are also possible, of course.

I now have a strong feeling that all of that has already been discussed somewhere. If that is really the case, please, point me to the discussion and I will try to join/revive it.

### CODE REVIEW

**Comment by rscohn2** [comment]
Thanks for the PR. I skimmed the refactoring and that looks ok. I need to run tests and take a deeper look at some of the changes--probably late next week.

**Comment by rscohn2** [comment]
Can you rebase to latest develop? My tests are failing because of an related upstream bug that has been corrected: https://github.com/rscohn2/oneapi-spack-tests/pull/34

**Comment by skosukhin** [comment]
@rscohn2 the branch is rebased to the latest `develop` now

**Comment by rscohn2** [comment]
> I now have a strong feeling that all of that has already been discussed somewhere. If that is really the case, please, point me to the discussion and I will try to join/revive it.

We never discussed ``icx -mkl``, but did discuss the general issue when first creating the packages. It wasn't in one single thread and you already understand all the issues so no need to go back there. It is convenient to simply add -mkl to the compile line and everything works, and we lost some of that.

Oneapi installer was implemented as a self-contained package manager. All components under a single prefix, components are optional, you can install multiple versions, environment variables to activate packages, ...

Spack had its own way of doing all this. Wherever possible, I tried to make oneapi do things the spack way. Decompose into the maximum number of packages as allowed by the installers. Components only in a single component. The original oneapi packages did not rely on the vars.sh files, but I gave up on that.

No solution is perfect. For example, I wanted to have the compiler runtimes to be a separate package so MKL would depend on that, but not the compiler. Intel decided to only provide linux compiler runtimes as apt/yum packages (not standalone installs) so that would have been difficult.

I feel like any solution that has the same components in multiple prefixes could lead to confusing and surprising conflicts for a spack user. That is why I am not in favor of installing mkl under the compiler prefix.

I am not against allowing variants or metapackages that create combinations of packages that work differently, but I don't have time to work on that.

**Review by rscohn2 — DISMISSED** [review]
The refactoring looks good.

RPATHing the internal compiler dependencies makes sense. It would be helpful if @alalazo or some other spack developer would confirm that the cfg file method is the most appropriate method.

I generated the cfg files by installing the compiler and inspected them, but did not do any testing, other than my normal oneapi package tests. I assume you verified that it does make the libraries accessible. Same for the mkl change. I don't use modules and have only done ad hoc testing.

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py [hunk 1]
```diff
@@ -3,13 +3,93 @@
...
+            'url': 'https://registrationcenter-download.intel.com/akdlm/irc_nas/17513/l_dpcpp-cpp-compiler_p_2021.1.2.63_offline.sh',
+            'sha256': '68d6cb638091990e578e358131c859f3bbbbfbf975c581fd0b4b4d36476d6f0a'
+        },
+        'ftn': {
+            'url': 'https://registrationcenter-download.intel.com/akdlm/irc_nas/17508/l_fortran-compiler_p_2021.1.2.62_offline.sh',
+            'sha256': '29345145268d08a59fa7eb6e58c7522768466dd98f6d9754540d1a0803596829'
+        }
+    }
+]
+
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:92]: This code, and what it replaces, appear to be the same length. I prefer the older approach.

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py [hunk 2]
```diff
@@ -157,8 +127,68 @@ def setup_run_environment(self, env):
...
+    @run_after('install')
+    def extend_config_flags(self):
+        # Extends compiler config files to inject additional compiler flags.
+        # Inject rpath flags to the runtime libraries. A pedantic implementation would
+        # use cc_rpath_arg, cxx_rpath_arg, f77_rpath_arg and fc_rpath_arg methods of
+        # spack.compilers.intel and spack.compilers.oneapi for the Classic and OneAPI
+        # compilers, respectively, but that looks like an overkill, given that all of
+        # them are the same. Also, it is unclear whether we should really
+        # use _ld_library_path because it looks like the only rpath that needs to be
+        # injected is self.component_prefix.linux.compiler.lib.intel64_lin.
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py]: > Also, it is unclear whether we should really use _ld_library_path...

Can you clarify this like:

> Also, it is unclear whether we should really use *all elements* of _ld_library_path...

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py [hunk 3]
```diff
@@ -67,41 +67,54 @@ class IntelOneapiMkl(IntelOneApiLibraryPackage):
...
+        # flag of the Intel compilers are used outside of the Spack build environment.
+        # We should not try to take care of other compilers because the users have to
+        # provide the linker flags anyway and are expected to take care of the RPATHs
+        # flags too. We prefer the POST flags over the PRE ones to allow for overriding
+        # of the injected RPATHs. Also, we use append_path instead of append_flags
+        # because the latter is not yet supported in the default modulefile templates.
+        for d in self._find_mkl_libs('+shared' in self.spec).directories:
+            flag = '-Wl,-rpath,{0}'.format(d)
+            env.append_path('__INTEL_POST_CFLAGS', flag, separator=' ')
+            env.append_path('__INTEL_POST_FFLAGS', flag, separator=' ')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py:106]: If you want to override, why use `append_path` (or maybe I misunderstood who is overriding who here)? `env.set` would be idiomatic (it wouldn't apply if if you wanted to combine flags, but based on your comment, it sounds like if that was desired you could do `env.set` on the PRE flags).

I also have a question: when you say the "Intel compilers" do you mean `%intel`? I assume it's not for `%oneapi` (since this PR updates the oneAPI compiler binaries to insert these RPATHs).

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py [hunk 4]
```diff
@@ -157,8 +127,68 @@ def setup_run_environment(self, env):
...
+                patchelf(file, fail_on_error=False)
+
+    @run_after('install')
+    def extend_config_flags(self):
+        # Extends compiler config files to inject additional compiler flags.
+        # Inject rpath flags to the runtime libraries. A pedantic implementation would
+        # use cc_rpath_arg, cxx_rpath_arg, f77_rpath_arg and fc_rpath_arg methods of
+        # spack.compilers.intel and spack.compilers.oneapi for the Classic and OneAPI
+        # compilers, respectively, but that looks like an overkill, given that all of
+        # them are the same. Also, it is unclear whether we should really
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py]: I think it's worth the extra lines to be generic about this:

```
cspec = 'oneapi@{0}'.format(self.spec.version)
compiler = spack.compiler.Oneapi(cspec, self.spec.arch.os, self.spec.arch.target)
bin_rpath_pairs = [
  ('icx', compiler.cc_rpath_arg),
  ...
]
for ...
  flags = ' '.join(['{0},{1}'.format(d) for d in self._ld_library_path()])
```

Alternatively, I think rephrasing:

```        
# Inject rpath flags to the runtime libraries. A pedantic implementation would
# use cc_rpath_arg, cxx_rpath_arg, f77_rpath_arg and fc_rpath_arg methods of
# spack.compilers.intel and spack.compilers.oneapi for the Classic and OneAPI
# compilers, respectively, but that looks like an overkill, given that all of
# them are the same. Also, it is unclear whether we should really
```

as 

```
TODO: this uses a static string for the rpath argument, but should actually
make sure that it matches the cc_rpath_arg etc. arguments defined in
spack.compilers.oneapi (for now, these are inherited from spack.compilers.compiler.Compiler):
these can theoretically be different for different compiler versions and for
different languages (C, C++, and Fortran), but in practice are not.
```

could also be sufficient (EDIT: I forgot to add this line)

Overall it's kind of annoying to get the code to do this idiomatically WRT how things are set up in the compilers, but it also took me a while to understand this comment, so I'm trying to figure out how to simplify that.

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a couple questions and a couple requests

**Comment by scheibelp** [comment]
Oh and I appreciated the note in the PR description about how most of these changes are refactoring, but I wanted to check precisely. My understanding is that changes in the following files are purely refactoring (no logic changes):

- intel-oneapi-mpi
- intel-oneapi-dpl
- intel-oneapi-dnn
- extrae
- build_systems/oneapi.py
- Everything in packages/intel-onapi-compiler/package.py except:
    - inject_rpaths
    - extend_config_flags
- Everything in intel-oneapi-mkl/package.py except
    - setup_run_environment

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py (hunk 1 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:92]: Same length but makes it easier to read that:
1. It does not really matter what is the package archive (the shell scripts are archives essentially) and what is the resource archive: they are semantically equal: one is C/C++ and another one is Fortran.
2. Each version of the package is represented by a pair of archives of the same version and there are no overlaps (a `resource` belongs to one version only).

I also think it will be easier to maintain the list when we add support for other platforms.

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py (hunk 3 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py:106]: 1. There are `__INTEL_PRE_CFLAGS` and `__INTEL_POST_FFLAGS` (see [here](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compilation/supported-environment-variables.html)). In the comment, I tried to explain why we should use the latter.
2. We should not `env.set` because the user might have `__INTEL_POST_FFLAGS` set to something already before loading the module and we want just to extend the value.
3. I mean both `%intel` and `%oneapi` and the modifications to `intel-oneapi-compilers` do not set these RPATHs because `intel-oneapi-compilers` is a separate package and doesn't know where `intel-oneapi-mkl` is installed (if at all).

Sorry for the confusion.

File: var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py (hunk 4 above)
Comment (skosukhin) [reply var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py]: Thanks for the hint. I didn't know how to create an instance of the compiler class properly in this case. We, however, would need to do it for `%intel` as well (`intel-oneapi-compilers` provides both `%intel` and `%oneapi`), therefore, I will just rephrase the comment.

**Comment by skosukhin** [comment]
@scheibelp Again, sorry for this. I will try to avoid putting so much stuff in on PR in the future. It's all related but could be really split in several different ones.

- `intel-oneapi-mpi` - refactoring plus:
    - avoid filtering the wrappers twice: replace both `I_MPI_SUBSTITUTE_INSTALLDIR` and `__EXEC_PREFIX_TO_BE_FILLED_AT_INSTALL_TIME__` in one go;
- `intel-oneapi-dpl` - address the renaming of `component_path` only;
- `intel-oneapi-dnn` - address the renaming of `component_path` only;
- `extrae` - address the renaming of `component_path` only;
- `build_systems/oneapi.py` - refactoring;
- `intel-oneapi-compilers` - refactoring plus:
    - do not inject RPATHs to directories without shared library files;
    - do not call `patchelf` several times on the same file:
        - `self.component_prefix.linux.compiler.lib.intel64` is a symlink to `self.component_prefix.linux.compiler.lib.intel64_lin` - no need to put the former on the list;
        - `icx-lto.so` gets patched as a file residing in `self.component_prefix.linux.lib` - no need to put it on the list explicitly;
    -  inject RPATHs to the runtime libraries into the produced executables;
- `intel-oneapi-mkl` - refactoring plus:
    - `setup_run_environment`.

File: var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py (hunk 3 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py:106]: Oh I think I see: here you mean override in the sense that if the user places other RPATHs anywhere in the options, that they will come before the ones inserted with this mechanism, and therefore take precedence. Can you clarify that in the comment?

Also, according to https://lmod.readthedocs.io/en/latest/051_tcl_modulefiles.html, TCL at least doesn't appear to have an "append-flags" function (the same appears to be true for [lmod](https://lmod.readthedocs.io/en/latest/050_lua_modulefiles.html)):

* If they in fact don't, can you rephrase the comment (if this is the case, then I think you did it the way the would have intended and there is no problem with this)?
* If they do, can you move this part of the comment to a TODO comment specifically on these`append_path` calls?

Comment (skosukhin) [reply var/spack/repos/builtin/packages/intel-oneapi-mkl/package.py:106]: @scheibelp I have updated the comment related to the choice of `POST` flags and removed the one about `append_path`. What I actually meant in the latter was that we could add support for `append_flags` in the module file generation via the `append-path` function with the space as delimiter.

**Comment by scheibelp** [comment]
(FYI I can look at this again on Monday 6/27)

**Comment by scheibelp** [comment]
It sounds like https://github.com/spack/spack/pull/30981#issuecomment-1158305398 and https://github.com/spack/spack/pull/30981#issuecomment-1158594120 are the same except your summary adds:

> avoid filtering the wrappers twice: replace both I_MPI_SUBSTITUTE_INSTALLDIR and __EXEC_PREFIX_TO_BE_FILLED_AT_INSTALL_TIME__ in one go;

Is this a functional change? I'd assume doing it twice would not cause an error (so not doing it twice would be refactoring)

**Comment by skosukhin** [comment]
@scheibelp the comments are essentially the same, yes. I just tried to be as explicit as possible about all the changes. The rest depends on flavours in personal definitions of `cleaning`, `refactoring` and  `optimization`, I guess.

**Comment by scheibelp** [comment]
@spackbot run pipeline

**Comment by scheibelp** [comment]
Close/reopen to restart tests

(I thought that spackbot prompt would restart the whole CI but I was wrong)

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Comment by haampie** [comment]
Hm, this is should be improved... These linker flags `-Wl,-rpath` are passed in pre-processor `cc -E` and compilation `cc -c` mode, which causes `-Wunused-command-line-argument` to be triggered. This in turn breaks certain configure scripts that don't expect the compiler to dump to stdout/stderr (for example `cmake %oneapi@2021.4.0` fails during the bootstrap stage).

Those rpaths can better be set in `SPACK_SOME_VARIABLE_FOR_RPATHS` (name escapes me) so that Spack can add the flags depending on whether it's in `ccld` mode.

---

Curiously the behavior has changed over time... 2022.2.0 is not issuing a warning, 2021.4.0 is. In any case, it doesn't really make sense to pass linker flags to a compiler that's not going to link something.

**Comment by haampie** [comment]
Maybe easiest to just add `-Wno-unused-command-line-argument` for older versions of `intel-oneapi-compilers` then?

**Comment by skosukhin** [comment]
@haampie I wanted the config files for users who work with the compiler directly, outside of Spack. I agree that it makes sense to add `-Wno-unused-command-line-argument` for versions of the compiler that emit the warning.

**Comment by balay** [comment]
> I agree that it makes sense to add -Wno-unused-command-line-argument for versions of the compiler that emit the warning.

Will note: adding `-Wno-unused-command-line-argument` doesn't really work for me - as we need compilers to:
- give warnings/errors for options that are not compatible.
-  not give warnings/errors for options that are compatible
configure tests rely on this behavior for internal tests to be accurate. [if not - it results in broken configure tests - consequently broken builds]
